File: leetcode/sliding-window/0438_find_all_anagrams.cpp
// LC 438. Find All Anagrams in a String | Medium
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if (p.size() > s.size()) return {};
        vector<int> res, pc(26, 0), sc(26, 0);
        for (int i = 0; i < p.size(); i++) { pc[p[i]-'a']++; sc[s[i]-'a']++; }
        if (pc == sc) res.push_back(0);
        for (int i = p.size(); i < s.size(); i++) {
            sc[s[i]-'a']++; sc[s[i-p.size()]-'a']--;
            if (pc == sc) res.push_back(i - p.size() + 1);
        }
        return res;
    }
};

File: leetcode/sliding-window/0438_find_all_anagrams.py
# LC 438. Find All Anagrams in a String | Medium
from collections import Counter

class Solution:
    def findAnagrams(self, s, p):
        if len(p) > len(s): return []
        pc, sc = Counter(p), Counter(s[:len(p)])
        res = [0] if sc == pc else []
        for i in range(len(p), len(s)):
            sc[s[i]] += 1
            sc[s[i - len(p)]] -= 1
            if sc[s[i - len(p)]] == 0: del sc[s[i - len(p)]]
            if sc == pc: res.append(i - len(p) + 1)
        return res
